TypeCobol/Compiler/Diagnostics/CrossChecker.cs

```diff
@@ -226,81 +228,202 @@ public override bool Visit(Move move)
+                            if (keys == null || keys.Length == 0)
+                            {
+                                DiagnosticUtils.AddError(search, $"Cannot use binary SEARCH on '{searchedTable.Name}' because it has no KEY.");
+                            }
+
+                            //Main table and all parent tables must have at least one index
+                            if (parentTableDefinitions != null)
+                            {
+                                foreach (var table in parentTableDefinitions)
+                                {
+                                    var indexes = table.GetIndexes();
+                                    if (indexes == null || indexes.Length == 0)
+                                    {
+                                        var message = table == searchedTable
+                                            ? $"Cannot use binary SEARCH on '{searchedTable.Name}' because it is not indexed."
+                                            : $"Cannot use binary SEARCH on '{searchedTable.Name}' because parent table '{table.Name}' is not indexed.";
+                                        DiagnosticUtils.AddError(search, message);
+                                    }
+                                }
+                            }
+                            //else TC not supported
+                        }
+                    }
+                    else
+                    {
+                        //Not a table
+                        DiagnosticUtils.AddError(search, $"Cannot SEARCH in '{searchedTable.Name}', data item is not a table.");
+                    }
                 }
+                //else undefined reference
             }
+            //else it's a syntax error
+
             return true;
         }
 
-        /// <summary>
-        /// Visit a WhenSearch and set a flag to true if either:
-        /// WHEN does not use one of the declared keys (ascending or descending)
-        /// WHEN does not use the first index declared for the table when subscripting
-        /// </summary>
-        private class WhenSearchVisitor : AbstractAstVisitor
+        public override bool Visit(WhenSearch whenSearch)
         {
-            private readonly WhenSearch _whenSearch;
-            public bool IsInError { get; private set; } = false;
-
+            System.Diagnostics.Debug.Assert(whenSearch.Parent is Search);
+            var search = (Search) whenSearch.Parent;
 
-            public WhenSearchVisitor(WhenSearch whenSearch)
+            if (search.CodeElement.StatementType == StatementType.SearchBinaryStatement && _searchTables.TryGetValue(search, out var tableDefinitions))
             {
-                _whenSearch = whenSearch;
-            }
+                //TC not supported
+                if (tableDefinitions == null) return true;
 
+                //Main table
+                System.Diagnostics.Debug.Assert(tableDefinitions.Count > 0);
```

Comment:
It is `true` because we start searching from a table and the method includes it in its result. I don't understand the second part of your comment: the summary is written independently of the actual usages of the method, it refers to the behavior of the method but does not depend on how it is used.

Reply:
To be more precised, we store the parent table list only if the target of the search is actually a table so it means that in the dictionary, the list is either `null` (for TC) or not empty.

Reply:
Ok, I understand: how the dictionnary is filled implies that the list is null or contains at least an element.